lxd/storage/drivers/driver_dir_volumes.go

```diff
@@ -104,88 +80,31 @@ func (d *dir) CreateVolume(vol Volume, filler *VolumeFiller, op *operations.Oper
 
 // CreateVolumeFromBackup restores a backup tarball onto the storage device.
 func (d *dir) CreateVolumeFromBackup(vol Volume, snapshots []string, srcData io.ReadSeeker, optimizedStorage bool, op *operations.Operation) (func(vol Volume) error, func(), error) {
-	revert := true
-	revertPaths := []string{}
-
-	// Define a revert function that will be used both to revert if an error occurs inside this
-	// function but also return it for use from the calling functions if no error internally.
-	revertHook := func() {
-		for _, revertPath := range revertPaths {
-			os.RemoveAll(revertPath)
-		}
-	}
-
-	// Only execute the revert function if we have had an error internally and revert is true.
-	defer func() {
-		if revert {
-			revertHook()
-		}
-	}()
-
-	volPath := vol.MountPath()
-	err := vol.EnsureMountPath()
-	if err != nil {
-		return nil, nil, err
-	}
-	revertPaths = append(revertPaths, volPath)
-
-	// Find the compression algorithm used for backup source data.
-	srcData.Seek(0, 0)
-	tarArgs, _, _, err := shared.DetectCompressionFile(srcData)
+	// Run the generic backup unpacker
+	postHook, revertHook, err := genericBackupUnpack(d.withoutGetVolID(), d.name, vol, snapshots, srcData, op)
 	if err != nil {
 		return nil, nil, err
 	}
 
-	// Prepare tar extraction arguments.
-	args := append(tarArgs, []string{
-		"-",
-		"--strip-components=2",
-		"--xattrs-include=*",
-		"-C", volPath, "backup/container",
-	}...)
-
-	// Extract instance.
-	srcData.Seek(0, 0)
-	err = shared.RunCommandWithFds(srcData, nil, "tar", args...)
-	if err != nil {
-		return nil, nil, err
-	}
-
-	if len(snapshots) > 0 {
-		// Create new snapshots directory.
-		err := createParentSnapshotDirIfMissing(d.name, vol.volType, vol.name)
-		if err != nil {
-			return nil, nil, err
+	// Define a post hook function that can be run once the backup config has been restored.
+	// This will setup the quota using the restored config.
+	postHookWrapper := func(vol Volume) error {
+		if postHook != nil {
+			err := postHook(vol)
```

Comment:
@stgraber related to the genericBackupUnpack() function change, I'm not sure this postHook function is ever returned.

Reply:
It is not currently but as above, it may be in the future and I'd like to keep the number of return values to match that of CreateVolumeFromBackup.